service/resource/endpoint/delete.go

```diff
@@ -53,18 +24,22 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteState inter
 }
 
 func (r *Resource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
-	delete, err := r.newDeleteChange(ctx, obj, currentState, desiredState)
+	k8sEndpoint, err := r.newDeleteChange(ctx, obj, currentState, desiredState)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
 	patch := framework.NewPatch()
-	patch.SetDeleteChange(delete)
+	if isEmptyEndpoint(*k8sEndpoint) {
+		patch.SetDeleteChange(k8sEndpoint)
+	} else {
+		patch.SetUpdateChange(k8sEndpoint)
+	}
```

Comment:
This logic does not belong here. All the `New*Patch` methods should do is call the private functions and set the returned changes to the patch. That is it. Nothing more. In case there shouldn't be a delete change added to the patch, the according private method should just return nil. The same applies to the update change computation. Just move the condition to the private methods.

Reply:
> Just move the condition to the private methods.

I don't think it's enough. Depending on a condition `SetDeleteChange` or `SetUpdateChange` is called.

This condition can be moved to ApplyDeleteChange, but then the logic from ApplyUpdateChange has to be duplicated. I personally like the approach here. It feels clean.

Reply:
I tried to move things a few times here, nothing I tried felt better than the solution with this condition here. 
One idea was to move this inside the private functions which then return `nil` if there is no patch which needs to be applied. This just replaced this condition with one that checked for nil and more logic inside the private functions. It did not feel like I was gaining anything here.

Another attempt was to simply emit more framework changes no matter what and make sure that no side effects took place (similar to what we do here https://github.com/giantswarm/kvm-operator/blob/bf5c288c5610572bcf3f096b592c63821e7282fe/service/kvmconfig/v3/resource/deployment/update.go#L58 ). I faced the problem that we currently hard replace values inside an Endpoint with our desired state, emitting more changes lead to more unnecessary overwrites which felt unclean to me. 

It feels to me like this is my favorite way to handling this right now, I would be glad if we can disagree and commit on this. I will then do some more testing of an installation.

Reply:
> One idea was to move this inside the private functions which then return nil if there is no patch which needs to be applied. This just replaced this condition with one that checked for nil and more logic inside the private functions. It did not feel like I was gaining anything here.

This is what I want. The `New*Patch` methods should be very dumb and always do the same without any logic or condition. All the custom magic should happen in the private functions. This is the whole point of this discussion IIRC and this is also why getting this done should be a matter of 5 minutes. 

> Another attempt was to simply emit more framework changes

I do not know what this means. 

> I would be glad if we can disagree and commit on this

No. I think it is important to get on the same page. It is not like we know what we are talking about and can move on regardless. We should come to an agreement we can all live with but this includes understanding what we are doing. You should only disagree and commit if you know what you are talking about and not use it as last resort because of missing knowledge.

Reply:
I changed this around, `New*Patch` is now very dumb and the conditions are checked again in the Apply methods to be safe. Please take another look.